File: cars-backend/src/main/kotlin/com/cars/carsbackend/service/CarService.kt
package com.cars.carsbackend.service

import com.cars.carsbackend.dto.CarMapper
import com.cars.carsbackend.dto.CarRequest
import com.cars.carsbackend.dto.CarResponse
import com.cars.carsbackend.repository.CarRepository
import org.springframework.data.repository.findByIdOrNull
import org.springframework.stereotype.Service

@Service
class CarService(
    private val carRepository: CarRepository,
    private val mapper: CarMapper,
) {

    fun getAll() : List<CarResponse> {
        return carRepository.findAll()
            .map { mapper.convertToResponse(it)!! }
            .toList()
    }

    fun getById(id: Long): CarResponse? {
        return mapper.convertToResponse(
            car = carRepository.findByIdOrNull(id)
        )
    }

    fun post(car: CarRequest) {
        carRepository.save(mapper.convertToCar(car))
    }

    fun update(id: Long, updated: CarRequest): Boolean {
        val car = carRepository.findById(id)
        if (car.isEmpty) return false
        carRepository.save(car.get().copy(
            name = updated.name ?: car.get().name,
            engine = updated.engine ?: car.get().engine
        ))
        return true
    }

    fun delete(id: Long) {
        carRepository.deleteById(id)
    }
}

File: cars-frontend/src/CarService.js
import axios from "axios";

class CarService {

    baseUrl = "http://localhost:8080/api/cars"

    getAll = async () => {
        const response = await axios.get(this.baseUrl)
        return response.data
    }

    getById = async (id) => {
        const response = await axios.get(this.baseUrl + `/${id}`)
        return response.data
    }

    post = async (car) => {
        await axios.post(this.baseUrl, car)
    }

    update = async (id, updated) => {
        await axios.put(`${this.baseUrl}/${id}`, updated)
    }

    delete = async (id) => {
        await axios.delete(`${this.baseUrl}/${id}`)
    }
}

export default new CarService();

